src/MongrationDotNet/MigrationRunner.cs

```diff
@@ -3,97 +3,194 @@
+                            break;
+                        }
+                        finally
+                        {
+                            if(migrationApplied != null)
+                                migrationsApplied.Add(migrationApplied);
+                        }
                     }
-                    catch (Exception ex)
+                    else
                     {
-                        await SetMigrationAsErrored(migration);
-                        logger?.LogError(LoggingEvents.MigrationFailed,
-                            "Migration failed for type: {type}, version: {version} and description: {description} with exception: {exception}. Skipping other migrations.",
-                            migration.Type, migration.Version.ToString(), migration.Description, ex.Message);
-                        break;
+                        logger?.LogInformation(LoggingEvents.MigrationSkipped,
+                            "Migration has already been applied or in progress. Skipped migration for type: {type}, version: {version} and description: {description} ",
+                            migration.Type, migration.Version.ToString(),
+                            migration.Description);
+                    }   
+                }
+
+                return migrationsApplied;
+            }
+            catch (Exception ex)
+            {
+                logger?.LogError(LoggingEvents.MigrationFailed,
+                    "Migration failed with exception: {exception}", ex.Message);
+                throw;
+            }
+        }
+
+        private async Task<bool> AnyMigrationInProgress()
+        {
+            var count = await migrationDetailsCollection
+                .CountDocumentsAsync(x => x.Status == MigrationStatus.InProgress);
+            return count > 0;
+        }
+
+       /// <summary>
+       /// Sets migration is progress
+       /// If any other node is running the migration the calling node will poll till migration is completed or till timeout 
+       /// </summary>
+       /// <param name="migration">Migration to start</param>
+       /// <returns></returns>
+        private async Task<bool> SetMigrationInProgress(IMigration migration)
+        {
+            try
+            {
+                var setToProgressSuccessfully = false;
+                //if migration is set to rerun after previous run
+                if (migration.RerunMigration)
+                {
+                    DeleteResult result = null;
+                    var previousRun = await migrationDetailsCollection.Find(x => x.Version == migration.Version).SingleOrDefaultAsync();
+                    //delete the previous run if exits
+                    if(previousRun != null)
+                        result = await migrationDetailsCollection.DeleteOneAsync(x => x.Version == migration.Version);
+                    //if no previous run or deleted successfully, insert new details
```

Comment:
Sorry, this still introduces more concurrecy issues - if another node deleted and started the migration this code will also delete it and run it again. You have to use the database as the a transactional co-ordination point for every decision on what to do, and check that the state matches what you recieved when you made a decision, in the same way you did with the insert by id, so you know if another node has done the same change already.